cmd/internal/labeler/labeler.go

```diff
@@ -455,8 +467,92 @@ func createNumaNodeMappingLabel(mapping map[int][]string) string {
 	return strings.Join(parts, "_")
 }
 
-func (l *labeler) printLabels() {
+func (l *labeler) printLabelsToFile(labelFile string) error {
+	f, err := os.OpenFile(labelFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open file (%s): %w", labelFile, err)
+	}
+
+	defer f.Close()
+
+	for key, val := range l.labels {
+		if _, err := f.WriteString(key + "=" + val + "\n"); err != nil {
+			return fmt.Errorf("failed to write label (%s=%s) to file: %w", key, val, err)
+		}
+	}
+
+	return nil
+}
+
+func CreateAndPrintLabels(sysfsDRMDir, debugfsDRIDir string) {
+	l := newLabeler(sysfsDRMDir, debugfsDRIDir)
+
+	if err := l.createLabels(); err != nil {
+		klog.Warningf("failed to create labels: %+v", err)
+
+		return
+	}
+
 	for key, val := range l.labels {
 		fmt.Println(key + "=" + val)
 	}
 }
+
+// Gathers node's GPU labels on channel trigger or timeout, and write them to a file.
+// The created label file is deleted on exit (process dying).
+func Run(sysfsDrmDir, debugfsDRIDir, nfdFeatureFile string, updateInterval time.Duration, scanResources chan bool) {
+	l := newLabeler(sysfsDrmDir, debugfsDRIDir)
+
+	interruptChan := make(chan os.Signal, 1)
+	signal.Notify(interruptChan, syscall.SIGTERM, syscall.SIGINT, syscall.SIGHUP, syscall.SIGQUIT)
+
+	klog.V(1).Info("Starting GPU labeler")
+
+	running := true
+
+	for running {
+		timeout := time.After(updateInterval)
+
+		select {
+		case <-timeout:
+		case <-scanResources:
+		case interrupt := <-interruptChan:
+			klog.V(2).Infof("Interrupt %d received", interrupt)
+
+			running = false
+
+			continue
+		}
+
+		klog.V(1).Info("Ext resources scanning")
+
+		err := l.createLabels()
+		if err != nil {
+			klog.Warningf("label creation failed: %+v", err)
+
+			continue
+		}
+
+		if l.labelsChanged {
+			klog.V(1).Info("Writing labels")
+
+			if err := l.printLabelsToFile(nfdFeatureFile); err != nil {
+				klog.Warningf("failed to write labels to file: %+v", err)
+			}
```

Comment:
Are label creation / writing issues expected to be temporary?

If not, wouldn't it be better to exit loop and terminate process with an error code, so that the issue is more visible (in pod restart counts)?

Reply:
My thought process was to keep labeler as sort of an optional thing. Same as with initcontainer, GPU plugin's functionality doesn't depend on the success of the initcontainer.

I can't think of temporary failures that would occur in typical use, probably there are something. And actually, in case of failure, the labelChanged result should be invalidated and writing should be tried again later. Otherwise any intermittent failure would result in no labels (unless the devices change).